// @snps/docs/dist/simple-docs.js
import { readFile, writeFile, mkdir } from 'node:fs/promises';
import { join } from 'node:path';
import { randomUUID } from 'node:crypto';
// Simple documentation service that showcases Synapse capabilities
export class SimpleDocumentationService {
    pages = new Map();
    examples = new Map();
    tutorials = new Map();
    constructor() {
        this.initializeContent();
    }
    initializeContent() {
        // Initialize with sample content
        this.addPage({
            id: 'getting-started',
            title: 'Getting Started with Synapse',
            slug: 'getting-started',
            content: this.getGettingStartedContent(),
            category: 'getting-started',
            order: 1,
            tags: ['introduction', 'setup', 'quick-start']
        });
        this.addPage({
            id: 'core',
            title: 'Core Framework',
            slug: 'core',
            content: this.getCoreContent(),
            category: 'core',
            order: 2,
            tags: ['server', 'routing', 'database', 'auth']
        });
        this.addPage({
            id: 'futuristic',
            title: 'Futuristic Features',
            slug: 'futuristic',
            content: this.getFuturisticContent(),
            category: 'futuristic',
            order: 3,
            tags: ['pwa', 'voice', 'webassembly', 'webrtc']
        });
        this.addExample({
            id: 'basic-server',
            title: 'Basic Server Setup',
            description: 'Create a simple HTTP server with Synapse',
            code: `import { Server } from '@snps/core';
import { Router } from '@snps/routing';

const server = new Server({ port: 3000 });
const router = new Router();

router.get('/', (req, res) => {
  res.writeHead(200, { 'Content-Type': 'text/html' });
  res.end('<h1>Hello from Synapse!</h1>');
});

server.useRouter(router);
await server.start();`,
            language: 'typescript',
            category: 'Core',
            package: '@snps/core',
            isRunnable: true,
            dependencies: ['@snps/core', '@snps/routing']
        });
        this.addExample({
            id: 'ai-integration',
            title: 'AI Integration with Web AI',
            description: 'Use Google Web AI for browser-based AI generation',
            code: `import { AIService } from '@snps/ai';

const ai = new AIService({
  webAI: { enableBrowserAI: true }
});

// Initialize Web AI
await ai.initializeWebAI('your-google-ai-key');

// Generate text using browser AI
const response = await ai.generateText(
  'Write a story about a robot learning to paint',
  'gemini-pro',
  { useWebAI: true }
);

console.log(response.content);`,
            language: 'typescript',
            category: 'AI',
            package: '@snps/ai',
            isRunnable: true,
            dependencies: ['@snps/ai']
        });
    }
    addPage(page) {
        this.pages.set(page.id, page);
    }
    addExample(example) {
        this.examples.set(example.id, example);
    }
    addTutorial(tutorial) {
        this.tutorials.set(tutorial.id, tutorial);
    }
    getPage(slug) {
        return Array.from(this.pages.values()).find(p => p.slug === slug);
    }
    getAllPages() {
        return Array.from(this.pages.values()).sort((a, b) => a.order - b.order);
    }
    getAllExamples() {
        return Array.from(this.examples.values());
    }
    getAllTutorials() {
        return Array.from(this.tutorials.values());
    }
    async generateHomePage() {
        const pages = this.getAllPages();
        const examples = this.getAllExamples();
        return `<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Synapse Framework Documentation</title>
    <style>
        body { font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, sans-serif; margin: 0; padding: 0; background: #f8fafc; }
        .container { max-width: 1200px; margin: 0 auto; padding: 40px 20px; }
        .header { text-align: center; margin-bottom: 60px; }
        .header h1 { font-size: 3em; margin-bottom: 20px; color: #2d3748; }
        .header p { font-size: 1.2em; color: #718096; max-width: 600px; margin: 0 auto; }
        .nav { display: grid; grid-template-columns: repeat(auto-fit, minmax(300px, 1fr)); gap: 30px; margin-bottom: 60px; }
        .nav-card { background: white; padding: 30px; border-radius: 12px; box-shadow: 0 4px 6px rgba(0,0,0,0.1); transition: transform 0.2s; }
        .nav-card:hover { transform: translateY(-4px); }
        .nav-card h3 { margin: 0 0 15px 0; color: #2d3748; font-size: 1.5em; }
        .nav-card p { margin: 0 0 20px 0; color: #718096; line-height: 1.6; }
        .nav-card a { color: #667eea; text-decoration: none; font-weight: 600; }
        .nav-card a:hover { text-decoration: underline; }
        .examples { background: white; padding: 40px; border-radius: 12px; box-shadow: 0 4px 6px rgba(0,0,0,0.1); margin-bottom: 40px; }
        .examples h2 { margin: 0 0 30px 0; color: #2d3748; }
        .example { margin-bottom: 30px; padding: 20px; border-left: 4px solid #667eea; background: #f7fafc; }
        .example h4 { margin: 0 0 10px 0; color: #2d3748; }
        .example p { margin: 0 0 15px 0; color: #718096; }
        .example pre { background: #2d3748; color: #e2e8f0; padding: 20px; border-radius: 8px; overflow-x: auto; margin: 15px 0; }
        .example .meta { display: flex; gap: 10px; margin-bottom: 15px; }
        .example .meta span { background: #e2e8f0; padding: 4px 8px; border-radius: 4px; font-size: 0.9em; color: #4a5568; }
    </style>
</head>
<body>
    <div class="container">
        <div class="header">
            <h1>🚀 Synapse Framework</h1>
            <p>Revolutionary TypeScript framework with zero dependencies, cutting-edge features, and enterprise-grade capabilities</p>
        </div>
        
        <div class="nav">
            ${pages.map(page => `
            <div class="nav-card">
                <h3>${page.title}</h3>
                <p>${page.content.substring(0, 150)}...</p>
                <a href="/${page.slug}">Read More →</a>
            </div>
            `).join('')}
        </div>
        
        <div class="examples">
            <h2>💻 Code Examples</h2>
            ${examples.map(example => `
            <div class="example">
                <h4>${example.title}</h4>
                <p>${example.description}</p>
                <div class="meta">
                    <span>${example.language}</span>
                    <span>${example.category}</span>
                    <span>${example.package}</span>
                </div>
                <pre><code>${example.code}</code></pre>
            </div>
            `).join('')}
        </div>
    </div>
</body>
</html>`;
    }
    async generatePage(slug) {
        const page = this.getPage(slug);
        if (!page) {
            return this.get404Page();
        }
        return `<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>${page.title} - Synapse Framework</title>
    <style>
        body { font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, sans-serif; margin: 0; padding: 0; background: #f8fafc; }
        .container { max-width: 800px; margin: 0 auto; padding: 40px 20px; }
        .content { background: white; padding: 40px; border-radius: 12px; box-shadow: 0 4px 6px rgba(0,0,0,0.1); }
        .content h1 { margin: 0 0 30px 0; color: #2d3748; }
        .content h2 { margin: 30px 0 20px 0; color: #4a5568; }
        .content h3 { margin: 25px 0 15px 0; color: #4a5568; }
        .content p { margin: 0 0 20px 0; color: #4a5568; line-height: 1.6; }
        .content pre { background: #f7fafc; padding: 20px; border-radius: 8px; overflow-x: auto; margin: 20px 0; }
        .content code { background: #f7fafc; padding: 2px 6px; border-radius: 4px; font-family: 'Monaco', 'Menlo', monospace; }
        .content ul, .content ol { margin: 0 0 20px 0; padding-left: 30px; }
        .content li { margin: 0 0 10px 0; color: #4a5568; line-height: 1.6; }
        .back-link { display: inline-block; margin-bottom: 20px; color: #667eea; text-decoration: none; }
        .back-link:hover { text-decoration: underline; }
    </style>
</head>
<body>
    <div class="container">
        <a href="/" class="back-link">← Back to Documentation</a>
        <div class="content">
            ${page.content}
        </div>
    </div>
</body>
</html>`;
    }
    get404Page() {
        return `<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Page Not Found - Synapse Framework</title>
    <style>
        body { font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, sans-serif; margin: 0; padding: 0; background: #f8fafc; display: flex; align-items: center; justify-content: center; min-height: 100vh; }
        .container { text-align: center; }
        .container h1 { font-size: 4em; margin-bottom: 20px; color: #2d3748; }
        .container p { font-size: 1.2em; color: #718096; margin-bottom: 30px; }
        .container a { color: #667eea; text-decoration: none; font-weight: 600; }
        .container a:hover { text-decoration: underline; }
    </style>
</head>
<body>
    <div class="container">
        <h1>404</h1>
        <p>Page not found</p>
        <a href="/">← Back to Documentation</a>
    </div>
</body>
</html>`;
    }
    getGettingStartedContent() {
        return `# Getting Started with Synapse

Welcome to the Synapse framework - a revolutionary TypeScript framework built with zero dependencies and cutting-edge features.

## What is Synapse?

Synapse is a comprehensive TypeScript framework that provides everything you need to build modern web applications. It's built with:

- **Zero Dependencies**: No external dependencies, just pure TypeScript and Node.js
- **100% Test Coverage**: Comprehensive testing across all packages
- **Type Safety**: Complete TypeScript support throughout
- **Modular Architecture**: 21 independent packages for different capabilities
- **Cutting-Edge Features**: AI, blockchain, PWA, voice interfaces, WebAssembly, WebRTC

## Quick Start

### 1. Install Synapse

\`\`\`bash
npm install @snps/core @snps/routing @snps/database
\`\`\`

### 2. Create Your First App

\`\`\`typescript
import { Server } from '@snps/core';
import { Router } from '@snps/routing';

const server = new Server({ port: 3000 });
const router = new Router();

router.get('/', (req, res) => {
  res.writeHead(200, { 'Content-Type': 'text/html' });
  res.end('<h1>Hello Synapse!</h1>');
});

server.useRouter(router);
await server.start();
\`\`\`

### 3. Run Your App

\`\`\`bash
npx tsx app.ts
\`\`\`

Visit http://localhost:3000 to see your app running!

## Next Steps

- Explore the [Core Framework](/core) documentation
- Check out [Futuristic Features](/futuristic) for cutting-edge capabilities
- Browse [Code Examples](/examples) for practical implementations

Ready to build the future? Let's get started!`;
    }
    getCoreContent() {
        return `# Core Framework

The Synapse core framework provides the foundation for all applications with HTTP server, routing, database, authentication, templating, and testing capabilities.

## Packages

### @snps/core
The heart of the Synapse framework with HTTP server, middleware, caching, performance monitoring, WebSocket support, logging, configuration management, error handling, and monitoring dashboard.

### @snps/routing
Advanced routing system with path matching, parameters, middleware support, and route grouping.

### @snps/database
In-memory database with ORM capabilities, QueryBuilder, Model class, relationships, and validation.

### @snps/auth
Complete authentication system with JWT, sessions, password hashing, OAuth2, CSRF protection, and security management.

### @snps/templating
Template engine for .webml files with variable substitution, conditional blocks, loops, and HTML escaping.

### @snps/testing
Comprehensive testing framework with test runner, mocks, spies, stubs, and advanced testing utilities.

## Getting Started with Core

### Basic Server Setup

\`\`\`typescript
import { Server } from '@snps/core';
import { Router } from '@snps/routing';

const server = new Server({
  port: 3000,
  enableCaching: true,
  enableSecurity: true,
  enablePerformanceMonitoring: true
});

const router = new Router();

router.get('/', (req, res) => {
  res.writeHead(200, { 'Content-Type': 'application/json' });
  res.end(JSON.stringify({ message: 'Hello from Synapse!' }));
});

server.useRouter(router);
await server.start();
\`\`\`

The core framework provides everything you need to build robust, scalable applications with TypeScript!`;
    }
    getFuturisticContent() {
        return `# Futuristic Features

Synapse includes revolutionary futuristic capabilities with Progressive Web Apps, voice interfaces, WebAssembly, and WebRTC for the next generation of web applications.

## Packages

### @snps/pwa
Progressive Web App support with service workers, offline functionality, push notifications, and native app-like installation.

### @snps/voice
Voice user interfaces with speech recognition, speech synthesis, voice commands, and conversational AI.

### @snps/webassembly
WebAssembly support for high-performance computing with near-native execution speeds and multi-language support.

### @snps/webrtc
WebRTC for real-time communication with video/audio streaming, screen sharing, and peer-to-peer data transfer.

## PWA Example

\`\`\`typescript
import { PWAService, PWAManifestBuilder } from '@snps/pwa';

const pwa = new PWAService(
  new PWAManifestBuilder()
    .name('My Synapse App')
    .shortName('SynapseApp')
    .description('A futuristic PWA built with Synapse')
    .startUrl('/')
    .display('standalone')
    .themeColor('#667eea')
    .build(),
  {
    cacheName: 'my-app-v1',
    cacheStrategy: 'staleWhileRevalidate',
    precacheUrls: ['/', '/dashboard', '/offline.html']
  }
);

await pwa.generateManifest();
await pwa.generateServiceWorker();
await pwa.registerServiceWorker();
\`\`\`

## Voice Interface Example

\`\`\`typescript
import { VoiceService } from '@snps/voice';

const voice = new VoiceService({
  language: 'en-US',
  continuous: true,
  confidence: 0.7
});

// Start voice listening
const session = await voice.startListening('user123');

// Add voice commands
await voice.addCommand('navigate to dashboard', 'navigate', { target: 'dashboard' });

// Speak text
await voice.speak('Welcome to the futuristic application!');
\`\`\`

These futuristic features represent the cutting edge of web development!`;
    }
    async generateStaticSite(outputDir) {
        await mkdir(outputDir, { recursive: true });
        // Generate home page
        const homePage = await this.generateHomePage();
        await writeFile(join(outputDir, 'index.html'), homePage);
        // Generate individual pages
        for (const page of this.getAllPages()) {
            const pageHtml = await this.generatePage(page.slug);
            await writeFile(join(outputDir, `${page.slug}.html`), pageHtml);
        }
        console.log(`Static site generated in ${outputDir}`);
    }
}
