// @snps/ui/dist/components/Card.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import React from 'react';
import { cn } from '../utils';
export const Card = ({ children, title, className, variant = 'default', hoverable = false, padding = 'md', ...props }) => {
    const baseClasses = 'bg-white rounded-lg border transition-all duration-200';
    const variantClasses = {
        default: 'border-gray-200 shadow-sm',
        elevated: 'border-gray-200 shadow-lg',
        outlined: 'border-gray-300 shadow-none'
    };
    const paddingClasses = {
        none: '',
        sm: 'p-4',
        md: 'p-6',
        lg: 'p-8'
    };
    const hoverClasses = hoverable ? 'hover:shadow-md hover:border-gray-300 cursor-pointer' : '';
    return (_jsxs("div", { className: cn(baseClasses, variantClasses[variant], paddingClasses[padding], hoverClasses, className), ...props, children: [title && (_jsx("h3", { className: "text-lg font-semibold text-gray-900 mb-4", children: title })), children] }));
};
export const CardHeader = ({ children, className, ...props }) => {
    return (_jsx("div", { className: cn('px-6 py-4 border-b border-gray-200', className), ...props, children: children }));
};
export const CardBody = ({ children, className, ...props }) => {
    return (_jsx("div", { className: cn('px-6 py-4', className), ...props, children: children }));
};
export const CardFooter = ({ children, className, ...props }) => {
    return (_jsx("div", { className: cn('px-6 py-4 border-t border-gray-200 bg-gray-50', className), ...props, children: children }));
};
export const CardTitle = ({ children, className, ...props }) => {
    return (_jsx("h3", { className: cn('text-lg font-semibold text-gray-900', className), ...props, children: children }));
};
